### PR TITLE
Serialize schema to a file

### PR DESCRIPTION
**Pull request by mansu**
Move `FieldDef` to `LuceneFieldDef` into metadata.proto. Add a new metadata proto file for storing schema. Add serialization code and related unit tests. Also refactored `SchemaAwareLogDocumentBuilderImpl.java. Update `LogBuilder` to use the schema. 

This code is part of #389. Refactored it so it's easier to review.

### CODE REVIEW

File: kaldb/src/main/java/com/slack/kaldb/logstore/DocumentBuilder.java [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
 
+import com.slack.kaldb.metadata.schema.LuceneFieldDef;
 import java.io.IOException;
+import java.util.Map;
 import org.apache.lucene.document.Document;
 
 public interface DocumentBuilder<T> {
   Document fromMessage(T message) throws IOException;
+
+  Map<String, LuceneFieldDef> getSchema();
```
Comment (vthacker) [inline kaldb/src/main/java/com/slack/kaldb/logstore/DocumentBuilder.java:11]: does this need to be public?

File: kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+ *
+ * <p>In case of a field conflict, this class uses FieldConflictPolicy to handle them.
+ *
+ * <p>NOTE: Currently, if building a document raises errors, we still register the type of the
+ * fields in this document partially. While the document may not be indexed, this partial field
+ * config will exist in the system. For now, we assume storing this metadata is fine since it is
+ * rarely an issue and helps with performance. If this is an issue, we need to scan the json twice
+ * to ensure document is good to index.
+ */
+public class SchemaAwareLogDocumentBuilderImpl implements DocumentBuilder<LogMessage> {
```
Comment (vthacker) [inline kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:38]: can we not change the package name in this PR? Or let's do it in a previous PR?

It's impossible to tell what changed in this class since the package was moved and also the number of lines here is much lesser.

So to make it a little more friendly to review can we wither list down the changes from the previous impl, or do the package change in a separate PR?

File: kaldb/src/main/java/com/slack/kaldb/logstore/DocumentBuilder.java (hunk 1 above)
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/logstore/DocumentBuilder.java:11]: Interfaces are public by default.

File: kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java (hunk 2 above)
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java]: Done.

**Comment by vthacker** [comment]
LuceneFieldDef proto has storeNumericDocValue, but it should just be storeDocValues (not numeric specific). We can add doc-values to string fields and that's the only way to do basic group by on a string field like hostname. Today we can't do that with the current design.

I'd like to see LuceneFieldDef take the field type enum instead of a string. That's looks safer to me. If not we should add validation in the ctor of LuceneFieldDef and some tests to validate it.

I think we should split TextField into two - How it's currently implemented is confusing and we can make it simpler by splitting it into two - TextField and StringField
- TextField is a field that can be indexed/stored/analyzed and never have doc values
- StringField can be indexed and have doc values but never be analyzed.

I like the idea of adding a metadata map to ChunkSchema however I'm curious to learn if you have an use-case in mind for it? Or is it just a good to have sort of thing?

Can we add some unit tests for SchemaAwareLogDocumentBuilderImpl#getJsonType that tests how it chooses LONG vs INT correctly?

Is the name SchemaAwareLogDocumentBuilderImpl is confusing? It's not per document right? 

Here is a test case that potentially exposes a bug? If I am understanding correctly Since the test is building both document using SchemaAwareLogDocumentBuilderImpl, the code should be able track that "intField" was indexed with a numeric value for document 1 and string value for document 2?

```
@Test
  public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
    SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
    assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
    assertThat(docBuilder.getSchema().size()).isEqualTo(17);

    LogMessage msg1 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "1",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message1",
                    "intField", 1));
    Document testDocument1 = docBuilder.fromMessage(msg1);
    assertThat(testDocument1.getFields().size()).isEqualTo(10);

    LogMessage msg2 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "2",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message2",
                            "intField", "not_an_int"));
    Document testDocument2 = docBuilder.fromMessage(msg2);
    // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
    // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
    assertThat(testDocument2.getFields().size()).isEqualTo(11);
  }
  ```

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+package com.slack.kaldb.metadata.schema;
+
+import com.google.common.base.Objects;
+import com.slack.kaldb.metadata.core.KaldbMetadata;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Map;
+
+/** Schema for a chunk. */
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:10]: This needs better class documentation around what the purpose of the metadata is for, who uses it, and why this class is needed.

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/FieldType.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+import org.apache.lucene.document.FloatDocValuesField;
+import org.apache.lucene.document.FloatPoint;
+import org.apache.lucene.document.IntPoint;
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.document.NumericDocValuesField;
+import org.apache.lucene.document.StoredField;
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.document.TextField;
+
+public enum FieldType {
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/metadata/schema/FieldType.java:17]: This enum should have documentation.

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      String name,
+      String fieldType,
+      boolean isStored,
+      boolean isIndexed,
+      boolean isAnalyzed,
+      boolean storeNumericDocValue) {
+    super(name);
+    this.fieldType = FieldType.valueOf(fieldType.trim().toUpperCase());
+    if (isAnalyzed && !isIndexed) {
+      throw new InvalidFieldDefException("Cannot set isAnalyzed without setting isIndexed");
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java:27]: This should be using `Preconditions` checks, (ie, `checkArgument`), as that is the pattern used for all of the other `KaldbMetadata` classes.

**Comment by mansu** [comment]
> LuceneFieldDef proto has storeNumericDocValue, but it should just be storeDocValues (not numeric specific). We can add doc-values to string fields and that's the only way to do basic group by on a string field like hostname. Today we can't do that with the current design.
Updated the variable to StoreDocValues.

> 
> I'd like to see LuceneFieldDef take the field type enum instead of a string. That's looks safer to me. If not we should add validation in the ctor of LuceneFieldDef and some tests to validate it.
>
This code would also power the config eventually. So, kept it as a string field. Added tests to make sure a invalid type thows exception. 
> I think we should split TextField into two - How it's currently implemented is confusing and we can make it simpler by splitting it into two - TextField and StringField

> TextField is a field that can be indexed/stored/analyzed and never have doc values
> StringField can be indexed and have doc values but never be analyzed.
Added new String field type. Kept text as a separate field.

> I like the idea of adding a metadata map to ChunkSchema however I'm curious to learn if you have an use-case in mind for it? Or is it just a good to have sort of thing?
For now, it will help us add more schema metadata without making code changes. 

> Can we add some unit tests for SchemaAwareLogDocumentBuilderImpl#getJsonType that tests how it chooses LONG vs INT correctly?
> 
In JSON, we can't have int's everything is a Long. 

> Is the name SchemaAwareLogDocumentBuilderImpl is confusing? It's not per document right?
> 
The fromMessage method builds a document out of a message. The name is inline with our other classes. 

> Here is a test case that potentially exposes a bug? If I am understanding correctly Since the test is building both document using SchemaAwareLogDocumentBuilderImpl, the code should be able track that "intField" was indexed with a numeric value for document 1 and string value for document 2?
> 
> @Test
>   public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
>     SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
>     assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
>     assertThat(docBuilder.getSchema().size()).isEqualTo(17);
> 
>     LogMessage msg1 =
>             new LogMessage(
>                     MessageUtil.TEST_DATASET_NAME,
>                     "INFO",
>                     "1",
>                     Map.of(
>                             LogMessage.ReservedField.TIMESTAMP.fieldName,
>                             MessageUtil.getCurrentLogDate(),
>                             LogMessage.ReservedField.MESSAGE.fieldName,
>                             "Test message1",
>                     "intField", 1));
>     Document testDocument1 = docBuilder.fromMessage(msg1);
>     assertThat(testDocument1.getFields().size()).isEqualTo(10);
> 
>     LogMessage msg2 =
>             new LogMessage(
>                     MessageUtil.TEST_DATASET_NAME,
>                     "INFO",
>                     "2",
>                     Map.of(
>                             LogMessage.ReservedField.TIMESTAMP.fieldName,
>                             MessageUtil.getCurrentLogDate(),
>                             LogMessage.ReservedField.MESSAGE.fieldName,
>                             "Test message2",
>                             "intField", "not_an_int"));
>     Document testDocument2 = docBuilder.fromMessage(msg2);
>     // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
>     // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
>     assertThat(testDocument2.getFields().size()).isEqualTo(11);
>   }
There is no bug in this code. Were you thinking about CONVERT_AND_DUPLICATE_FIELD? That is different than CONVERT_FIELD_VALUE which only converts the field value.

**Comment by vthacker** [comment]
```
@Test
  public void testSameFieldDifferentValueTypes() throws JsonProcessingException {
    SchemaAwareLogDocumentBuilderImpl docBuilder = build(CONVERT_FIELD_VALUE, meterRegistry);
    assertThat(docBuilder.getIndexFieldConflictPolicy()).isEqualTo(CONVERT_FIELD_VALUE);
    assertThat(docBuilder.getSchema().size()).isEqualTo(17);

    LogMessage msg1 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "1",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message1",
                            "intField", 1));
    Document testDocument1 = docBuilder.fromMessage(msg1);
    assertThat(testDocument1.getFields().size()).isEqualTo(10);

    LogMessage msg2 =
            new LogMessage(
                    MessageUtil.TEST_DATASET_NAME,
                    "INFO",
                    "2",
                    Map.of(
                            LogMessage.ReservedField.TIMESTAMP.fieldName,
                            MessageUtil.getCurrentLogDate(),
                            LogMessage.ReservedField.MESSAGE.fieldName,
                            "Test message2",
                            "intField", "not_an_int"));
    Document testDocument2 = docBuilder.fromMessage(msg2);
    // Potential bug? This shoudn't be 10 right? now the same intField should be converted to a text field? We use CONVERT_FIELD_VALUE in the test
    // Also if you look closely to testDocument3 it shows the intField is being indexed with value 0 as a pointField and DocValue Field
    assertThat(testDocument2.getFields().size()).isEqualTo(10);
  }
```

This test still fails - We should be getting 10 according to our chat but it's 12

**Review by vthacker — DISMISSED** [review]

File: kaldb/src/test/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImplTest.java [hunk 6]
```diff
@@ -7,16 +7,17 @@
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldConflictPolicy.CONVERT_FIELD_VALUE;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldConflictPolicy.DROP_FIELD;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldConflictPolicy.RAISE_ERROR;
-import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldType;
-import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.FieldType.convertFieldValue;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.build;
 import static com.slack.kaldb.logstore.schema.SchemaAwareLogDocumentBuilderImpl.makeNewFieldOfType;
+import static com.slack.kaldb.metadata.schema.FieldType.convertFieldValue;
```
Comment (bryanlb) [inline kaldb/src/test/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImplTest.java:12]: Now that this code has been moved the corresponding tests should be moved in the appropriate class as well. We're testing `FieldType` static methods in the `SchemaAwareLogDocumentBuilderImplTest.java` class, when the `SchemaAwareLogDocumentBuilderImpl` no longer contains this code.

File: kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java [hunk 7]
```diff
@@ -48,285 +42,55 @@ public class SchemaAwareLogDocumentBuilderImpl implements DocumentBuilder<LogMes
...
-          return ((Double) value).longValue();
-        }
-        if (toType == FieldType.FLOAT) {
-          return ((Double) value).floatValue();
-        }
-      }
-
-      return null;
-    }
+  private static void addTextField(
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:44]: Doesn't look like this was removed as mentioned.

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * This schema class enforces schema for a chunk. The schema is only written in indexer and on the
+ * cache node the schema is read only.
+ *
+ * <p>The fieldMap is a map to the field name and LuceneFieldDef that stores the field definitions.
+ * Since this field is written and read my multiple threads it is hardcoded as a ConcurrentHashMap.
+ * New fields are added to the fieldMap when it's written and read during query.
+ */
+public class ChunkSchema extends KaldbMetadata {
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:18]: We seem to be missing the corresponding `ChunkSchemaStore` for persisting this metadata, and the associated test.

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package com.slack.kaldb.metadata.schema;
+
+import com.google.common.base.Objects;
+import com.slack.kaldb.metadata.core.KaldbMetadata;
+
+/*
+ * LuceneFieldDef describes the configs that can be set on a lucene field. This config defines how a field is indexed.
+ */
+public class LuceneFieldDef extends KaldbMetadata {
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/metadata/schema/LuceneFieldDef.java:9]: We seem to be missing the corresponding metadata stores and associated tests.

File: kaldb/src/main/java/com/slack/kaldb/logstore/LogDocumentBuilderImpl.java [hunk 10]
```diff
@@ -370,4 +372,9 @@ public Document fromMessage(LogMessage message) throws JsonProcessingException {
     }
     return doc;
   }
+
+  @Override
+  public Map<String, LuceneFieldDef> getSchema() {
+    return Collections.emptyMap();
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/logstore/LogDocumentBuilderImpl.java:378]: This should have a comment explaining why an empty map is appropriate here.

File: kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java [hunk 11]
```diff
@@ -348,30 +135,29 @@ public enum FieldConflictPolicy {
     CONVERT_AND_DUPLICATE_FIELD
   }
 
-  private static final Map<FieldType, FieldDef> defaultPropDescriptionForType =
+  private static final String DUMMY_FIELD = "DUMMY_FIELD";
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java:138]: This needs a comment explaining why a dummy field is being used here for the default properties.

File: kaldb/src/main/proto/metadata.proto [hunk 12]
```diff
@@ -147,3 +147,29 @@ message DatasetPartitionMetadata {
...
+// Schema definitions.
+message LuceneFieldDef {
+  // Name of the field
+  string name = 1;
+
+  // Type of the field
+  string type = 2;
+
+  // Store the value as a Lucene StoredField.
+  bool isStored = 3;
```
Comment (bryanlb) [inline kaldb/src/main/proto/metadata.proto:160]: These fields aren't using the correct naming convention, which will fail linting - see https://github.com/slackhq/kaldb/pull/222

File: kaldb/src/main/proto/metadata.proto [hunk 13]
```diff
@@ -147,3 +147,29 @@ message DatasetPartitionMetadata {
...
+  // Store doc values for a field so it's more efficient for aggregations.
+  bool storeDocValue = 5;
+}
+
+// This proto documents the schema for a chunk.
+message ChunkSchema {
+  // Name of the chunk.
+  string name = 1;
+  // Field map for each field in the lucene index in the chunk.
+  map<string, LuceneFieldDef> fieldDefMap = 2;
```
Comment (bryanlb) [inline kaldb/src/main/proto/metadata.proto:172]: This probably shouldn't be suffixed map, since none of the other maps are suffixed map - even in the same class.

File: kaldb/src/main/java/com/slack/kaldb/logstore/LogDocumentBuilderImpl.java (hunk 10 above)
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/logstore/LogDocumentBuilderImpl.java:392]: Fixed.

File: kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/metadata/schema/ChunkSchema.java:18]: Add tests.

File: kaldb/src/main/proto/metadata.proto (hunk 13 above)
Comment (mansu) [reply kaldb/src/main/proto/metadata.proto:175]: Tried removing the map, but the code is less readable. This is a map from an app perspective for field definitions and not meant to reflect the type (which also happens to match). So leaving it as is.

File: kaldb/src/test/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImplTest.java (hunk 6 above)
Comment (mansu) [reply kaldb/src/test/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImplTest.java]: Moved.

**Review by bryanlb — APPROVED** [review]
